### PR TITLE
Sort MonoItems by span instead of DefIndex.

### PR DESCRIPTION
**Pull request by cjgillot**
The codegen tests rely on items being process in the same order as they appear in the file.

Instead of sorting them by `DefIndex`, we should just sort them by `def_span`.

cc #90317.

### CODE REVIEW

**Comment by pierwill** [comment]
Thanks for fixing this, @cjgillot! :)

**Review by tmiasko — APPROVED** [review]

File: compiler/rustc_middle/src/mir/mono.rs [hunk 1]
```diff
@@ -353,37 +353,35 @@ impl<'tcx> CodegenUnit<'tcx> {
...
-                            | InstanceDef::CloneShim(..) => None,
+            let span = match item {
+                MonoItem::Fn(ref instance) => {
+                    match instance.def {
+                        // We only want to take HirIds of user-defined
+                        // instances into account. The others don't matter for
+                        // the codegen tests and can even make item order
+                        // unstable.
+                        InstanceDef::Item(def) => {
+                            def.did.as_local().map(|def_id| tcx.def_span(def_id))
```
[the remark this answers is not in the document]
Comment (tmiasko) [reply compiler/rustc_middle/src/mir/mono.rs]: OK, so it looks like we also need to include `SyntaxContext` to obtain "natural" order.

In that case, one thing I don't understand, what allows us to rely on`SyntaxContext` order? Doesn't it suffer from the similar issue as `DefIndex` given that we remap IDs when loading them from on disk cache, and then sort based on their values?

[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_middle/src/mir/mono.rs]: Marking as blocked until we reach a strategy for `SyntaxContext`. Each additional use of its `PartialOrd` impl will make it even harder to remove.

**Pull request closed without merging**